proxy/src/outbound.rs

```diff
@@ -54,6 +54,60 @@ impl<B> Outbound<B> {
             bind_timeout,
         }
     }
+
+    const DEFAULT_PORT: Option<u16> = Some(80);
+
+    /// TODO: Return error when `HostAndPort::normalize()` fails.
+    /// TODO: Use scheme-appropriate-default
+    fn normalize(authority: &http::uri::Authority) -> Option<HostAndPort> {
+        HostAndPort::normalize(authority, Self::DEFAULT_PORT).ok()
+    }
+
+    /// Determines the logical host:port of the request.
+    ///
+    /// 1. If the parsed URI includes an authority, use that.
+    /// 2. Otherwise, try to load the authority from the `Host` header.
+    ///
+    /// The port is either parsed from the authority or a default of 80 is used.
+    fn host_port(req: &http::Request<B>) -> Option<HostAndPort> {
+        req.uri()
+            .authority_part()
+            .and_then(Self::normalize)
+            .or_else(|| {
+                h1::authority_from_host(req)
+                    .and_then(|h| Self::normalize(&h))
+            })
```

Comment:
It seems like we could do `req.uri().authority_part().or_else(|| h1::authority_from_req(req)).and_then(Self::normalize))` and that would be better, if we can. (I'm guessing there's some reason we can't.)

Reply:
To do this, we have to clone the authority. (because authority_part returns a ref, and we can't make the `or_else` block return a ref in your suggestion). We only need a ref, though, so preferred not to clone just to pass the ref.

Reply:
OK, sounds good!